### PR TITLE
docker: Update Base images

### LINKED ISSUE #1091: Each docker image should use a specific version for its base images SP 3
- [ ] I have searched the [issues](https://github.com/openkfw/TruBudget/issues) of this repository and believe that this is not a duplicate.

## Description 😯

Currently there are some dockerfiles which are not referencing specfic versions of case images.
A specific version would prevent Trubudget from using unexpected updated versions.

Docker images should also follow OWASP rules

## Your Environment 🌎

| Tech            | Version |
| --------------- | ------- |
| Frontend        | v1.?.?  |
| API             | v1.?.?  |
| Blockchain      | v1.?.?  |
| Export-Service  | v1.?.?  |
| Email-Service   | v1.?.?  |
| Storage-Service | v1.?.?  |

### LINKED ISSUE #1191: Messed up env vars
The refactoring of the environment vars in this commit are completely wrong and should be reverted:

https://github.com/openkfw/TruBudget/commit/3ae302e57ffcc3ed41b0ea69bbe9469b5532f37b

For example: MULTICHAIN_FEED_ENABLED is never used, neither in a Docker (compose) file, nor in a config file.

### PR DESCRIPTION
**Pull request by daniel-arnauer**
Update Docker images

https://sysdig.com/blog/dockerfile-best-practices/#1-1

Closes #1091
Closes #1191

### CODE REVIEW

File: blockchain/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM node:10.5-slim
+FROM --platform=linux/amd64 node:18.4-bullseye
```
Comment (daniel-arnauer) [inline blockchain/Dockerfile:1]: will this work on intel cpu?

File: blockchain/Dockerfile [hunk 2]
```diff
@@ -1,29 +1,31 @@
...
 
 WORKDIR /home/node
-
-COPY package*.json ./
-
+COPY package.json ./
+COPY package-lock.json ./
 RUN npm ci
+COPY ./src/ ./src/
+RUN chmod +x ./src/multichain-feed/multichain-feed
```
Comment (mayrmartin) [inline blockchain/Dockerfile:22]: why don't simply 

- create a `multichain` user
- `chown` the feed directory to the `multichain` user
- run the node wrapper with the `multichain` user
- change the node wrapper to start `multichain-deamon` as `root` if needed

File: blockchain/Dockerfile [hunk 3]
```diff
@@ -1,29 +1,31 @@
...
-RUN chmod +x src/multichain-feed/multichain-feed
 
 ARG BUILDTIMESTAMP=''
 ARG CI_COMMIT_SHA=''
-
 ENV BUILDTIMESTAMP ${BUILDTIMESTAMP}
 ENV CI_COMMIT_SHA ${CI_COMMIT_SHA}
 
-# Run Chain
+# Start as root user: Multichain needs root access to the file system
```
Comment (mayrmartin) [inline blockchain/Dockerfile:30]: You sure? Can not find anything about this in their documentation, maybe `chown`/`chmod` the multichain dir?

File: email-notification-service/Dockerfile [hunk 4]
```diff
@@ -1,13 +1,21 @@
...
+COPY package-lock.json ./
 RUN npm ci
 
 COPY src src/
 COPY tsconfig.json .
 
 RUN npm run build
 
+# Create user
+RUN adduser -S -D -h /home/node newuser
```
Comment (mayrmartin) [inline email-notification-service/Dockerfile:15]: Give the user a meaningful name

File: excel-export-service/Dockerfile [hunk 5]
```diff
@@ -1,13 +1,19 @@
...
+COPY package-lock.json ./
 RUN npm ci
-
 COPY src src/
 COPY tsconfig.json .
-
 RUN npm run build
 
+# Create user
+RUN adduser -S -D -h /home/node newuser
```
Comment (mayrmartin) [inline excel-export-service/Dockerfile:13]: Give the user a meaningful name

File: logging-service/Dockerfile [hunk 6]
```diff
@@ -1,11 +1,19 @@
...
 COPY serve.js serve.js
-COPY package*.json ./
-
+COPY package.json ./
+COPY package-lock.json ./
 
 RUN npm ci
 
+# Create user
+RUN adduser -S -D -h /home/node newuser
```
Comment (mayrmartin) [inline logging-service/Dockerfile:12]: Give the user a meaningful name

File: storage-service/Dockerfile [hunk 7]
```diff
@@ -1,13 +1,20 @@
...
+COPY package-lock.json ./
 RUN npm ci
-
 COPY src src/
 COPY tsconfig.json .
 
 RUN npm run build
 
+# Create user
+RUN adduser -S -D -h /home/node newuser
```
Comment (mayrmartin) [inline storage-service/Dockerfile:14]: Give the user a meaningful name

File: api/Dockerfile [hunk 8]
```diff
@@ -16,4 +17,12 @@ ARG CI_COMMIT_SHA=''
 ENV BUILDTIMESTAMP ${BUILDTIMESTAMP}
 ENV CI_COMMIT_SHA ${CI_COMMIT_SHA}
 
+# Create user
+RUN useradd -ms /bin/bash newuser
```
Comment (mayrmartin) [inline api/Dockerfile:21]: Give the user a meaningful name

**Review by mayrmartin — CHANGES_REQUESTED** [review]

File: blockchain/Dockerfile (hunk 2 above)
Comment (daniel-arnauer) [reply blockchain/Dockerfile:22]: sounds good, does not work

File: email-notification-service/Dockerfile (hunk 4 above)
Comment (daniel-arnauer) [reply email-notification-service/Dockerfile:15]: suggest one please. Newuser is a meaningful name for this purpose

File: blockchain/Dockerfile [hunk 9]
```diff
@@ -1,29 +1,31 @@
-FROM node:10.5-slim
+FROM node:18.4.0-alpine
+
+RUN deluser --remove-home node
```
[the remark this answers is not in the document]
Comment (daniel-arnauer) [reply blockchain/Dockerfile:3]: Because only the root user is used in this container

File: blockchain/Dockerfile (hunk 2 above)
Comment (mayrmartin) [reply blockchain/Dockerfile:22]: * forgot: you might also need to change the `spawn` command of the multichain for this not only the docker file

File: excel-export-service/src/excel.ts [hunk 10]
```diff
@@ -31,7 +31,7 @@ export async function writeXLSX(
       useStyles: true,
       useSharedStrings: true,
     };
-    const { userId } = jwtDecode(token);
+    const { userId }: any = jwtDecode(token);
```
[the remark this answers is not in the document]
Comment (daniel-arnauer) [reply excel-export-service/src/excel.ts]: done

File: excel-export-service/src/excel.ts [hunk 11]
```diff
@@ -1,6 +1,7 @@
 import { AxiosInstance } from "axios";
 import { CustomExpressResponse } from "./types";
 import jwtDecode from "jwt-decode";
+// import jwtDecode, { JwtPayload } from "jwt-decode";
```
[the remark this answers is not in the document]
Comment (daniel-arnauer) [reply excel-export-service/src/excel.ts]: done

**Review by mayrmartin — APPROVED** [review]

**Review by georgimld — APPROVED** [review]

**Pull request merged**